File: 2.0/WearOS/WatchSprings-2.0/app/src/main/java/copyright/markmorris/allrightsreserved/watchsprings/screens/models/ShopViewModel.kt
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

package copyright.markmorris.allrightsreserved.watchsprings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ShopViewModel : ViewModel() {
    private val _shopResponse = MutableStateFlow<ShopResponse?>(null)
    val shopResponse = _shopResponse.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading = _isLoading.asStateFlow()

    private val _selectedItem = MutableStateFlow<ShopItem?>(null)
    val selectedItem = _selectedItem.asStateFlow()

    fun selectItem(item: ShopItem) {
        _selectedItem.value = item
    }

    fun refreshShop(forceReload: Boolean = false) {
        if (!forceReload && _shopResponse.value != null) return

        viewModelScope.launch {
            _isLoading.value = true
            try {
                val result = MainActivity.fortniteAPI.getDailyShop()
                _shopResponse.value = result
            } catch (e: Exception) {
                println("Error fetching shop data: ${e.message}")
            } finally {
                _isLoading.value = false
            }
        }
    }
}

File: 2.0/watchOS/WatchSprings-2.0/WatchSprings Watch App/screens/models/ShopViewModel.swift
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

import Foundation

@Observable
class ShopViewModel
{
    var shopResponse: ShopResponse? = nil
    var selectedItem: ShopItem? = nil
    var isLoading: Bool = false

    func select(item: ShopItem)
    {
        selectedItem = item
    }

    func refreshShop()
    {
        guard shopResponse == nil else { return } // Already loaded
   
        isLoading = true

        Task {
            do {
                let result = try await FortniteApiClient.instance.getDailyShop()
                await MainActor.run {
                    self.shopResponse = result
                    self.isLoading = false
                    print("Shop items loaded: \(result.shop.count)")
                }
            } catch {
                await MainActor.run {
                    print("Error loading shop items: \(error.localizedDescription)")
                    self.isLoading = false
                }
            }
        }
    }
}
